### DISCUSSION TITLE
httpd error 256 root - after load php module in httpd.conf

### DISCUSSION THREAD

**Discussion by toloekka** [Homebrew/discussions#3525] [Getting started]
Hi :)

I have a brand new Silicon Mac and I try Homebrew since php no longer are Default in macOS.

I have installed Homebrew, apache from Homebrew and php from Homebrew (both php 8.1.9 and 7.4).

After installing Homebrew and then apache, I got «It works» in my browser so so far it seem okay.

After installing php 8.1.9 and configured the httpd.conf file, restarted httpd I get «httpd error 256» like you see under here I type «brew services list»:
```
_httpd   error  256 root      ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist
php     started    2msstudio ~/Library/LaunchAgents/homebrew.mxcl.php.plist
php@7.4 none_ 
```

And nothing show in the browser now. It say something like this «This Site isn not available now».

It is when I load the php module in the httpd file I get this error (I restated step by step):
LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "2msstudio"
( I added my Home named «2msstudio» in .zprofile)

I tried this (without «2msstudio»):
LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so

And this:
LoadModule php7_module /opt/homebrew/opt/php@7.4/lib/httpd/modules/libphp7.so "2msstudio"
          

Before adding the php module in the httpd file and type this «brew services list», it show no error:
```
2msstudio@Toms-Mac-Studio httpd % brew services list 
Name    Status  User      File
httpd   started root      ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist
php     started 2msstudio ~/Library/LaunchAgents/homebrew.mxcl.php.plist
php@7.4 none              
```

Hope someone can help me out :)

Kind Regards
Tom Løkka
Norway

**Comment by SMillerDev** [comment]
What does the httpd log say?

**Reply by toloekka** [reply]
Is this the correct place for the error log?
```
-rw-r--r--  1 2msstudio  admin  2313 Aug 13 12:30 access_log
-rw-r--r--  1 2msstudio  admin  2995 Aug 13 12:30 error_log
2msstudio@Toms-Mac-Studio httpd % pwd
/opt/homebrew/var/log/httpd
2msstudio@Toms-Mac-Studio httpd % 
```
I see this in that error_log:
```
[Sat Aug 13 11:25:36.013618 2022] [mpm_prefork:notice] [pid 5358] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 11:25:36.014287 2022] [core:notice] [pid 5358] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 11:46:13.769255 2022] [mpm_prefork:notice] [pid 5358] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:00:12.481699 2022] [mpm_prefork:notice] [pid 7292] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:00:12.482268 2022] [core:notice] [pid 7292] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:02:07.701022 2022] [mpm_prefork:notice] [pid 7292] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:12:32.431919 2022] [mpm_prefork:notice] [pid 8437] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:12:32.432859 2022] [core:notice] [pid 8437] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:16:41.148619 2022] [mpm_prefork:notice] [pid 8437] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:16:42.241572 2022] [mpm_prefork:notice] [pid 8784] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:16:42.242097 2022] [core:notice] [pid 8784] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:26:52.165250 2022] [mpm_prefork:notice] [pid 8784] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:26:53.310226 2022] [mpm_prefork:notice] [pid 9574] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:26:53.310808 2022] [core:notice] [pid 9574] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:27:49.322499 2022] [mpm_prefork:notice] [pid 9574] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:27:50.411637 2022] [mpm_prefork:notice] [pid 9738] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:27:50.412186 2022] [core:notice] [pid 9738] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:28:13.476940 2022] [mpm_prefork:notice] [pid 9738] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:28:14.562823 2022] [mpm_prefork:notice] [pid 9898] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:28:14.563274 2022] [core:notice] [pid 9898] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:28:46.617249 2022] [mpm_prefork:notice] [pid 9898] AH00169: caught SIGTERM, shutting down
[Sat Aug 13 12:28:47.708045 2022] [mpm_prefork:notice] [pid 10060] AH00163: Apache/2.4.54 (Unix) configured -- resuming normal operations
[Sat Aug 13 12:28:47.708553 2022] [core:notice] [pid 10060] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'
[Sat Aug 13 12:30:12.292016 2022] [mpm_prefork:notice] [pid 10060] AH00169: caught SIGTERM, shutting down
```

**Reply by toloekka** [reply]
> What does the httpd log say?

Are you there? :)

**Reply by SMillerDev** [reply]
Okay, so the whole command seems to be failing. What does `/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND` do?

**Reply by toloekka** [reply]
I am not sure. I followed a youtube video. 
How can I dissable /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND?

I will try to find the youtube video where I found this.

**Reply by toloekka** [reply]
Here is the log" of the installation of apache/httpd wher I used the FOREGROUND command:
```
2msstudio@Toms-Mac-Studio ~ % brew install httpd
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/services).
==> New Formulae
censys             cql-proxy          libvatek           lucky-commit       ouch

You have 1 outdated formula installed.
You can upgrade it with brew upgrade
or list it with brew outdated.

==> Downloading https://ghcr.io/v2/homebrew/core/httpd/manifests/2.4.54
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/httpd/blobs/sha256:629e972fa257879e1fa9c7ada3c
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:629e972fa2
######################################################################## 100.0%
==> Pouring httpd--2.4.54.arm64_monterey.bottle.tar.gz
==> Caveats
DocumentRoot is /opt/homebrew/var/www.

The default ports have been set in /opt/homebrew/etc/httpd/httpd.conf to 8080 and in
/opt/homebrew/etc/httpd/extra/httpd-ssl.conf to 8443 so that httpd can run without sudo.

To restart httpd after an upgrade:
  brew services restart httpd
Or, if you don't want/need a background service you can just run:
  /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
==> Summary
🍺  /opt/homebrew/Cellar/httpd/2.4.54: 1,662 files, 31.9MB
==> Running `brew cleanup httpd`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
2msstudio@Toms-Mac-Studio ~ % 
2msstudio@Toms-Mac-Studio ~ % 
2msstudio@Toms-Mac-Studio ~ % brew services start httpd
Bootstrap failed: 5: Input/output error
Try re-running the command as root for richer errors.
Error: Failure while executing; `/bin/launchctl bootstrap gui/501 /Users/2msstudio/Library/LaunchAgents/homebrew.mxcl.httpd.plist` exited with 5.
2msstudio@Toms-Mac-Studio ~ % 
```

**Reply by SMillerDev** [reply]
Where did you use the command?

**Reply by toloekka** [reply]
In Terminal

**Reply by SMillerDev** [reply]
Okay, but without the output of that command there isn't much I can do to help you.

**Reply by toloekka** [reply]
Okay. Then I can uninstall Homebrew and start over again.
But shall I not use the command "/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND" again?

Will this command /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)" uninstall every files from Homebrew?

**Reply by carlocab** [reply]
An easier place to start might be to show the output of running `/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND`. Please run `/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND`, and **not** `brew services start httpd`, which is what you did [above](https://github.com/orgs/Homebrew/discussions/3525#discussioncomment-3404542).

**Reply by toloekka** [reply]
```
2msstudio@Toms-Mac-Studio ~ % /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
httpd: Syntax error on line 182 of /opt/homebrew/etc/httpd/httpd.conf: Cannot load /usr/local/opt/php/lib/httpd/modules/libphp.so into server: dlopen(/usr/local/opt/php/lib/httpd/modules/libphp.so, 0x000A): tried: '/usr/local/opt/php/lib/httpd/modules/libphp.so' (no such file)
```

**Reply by SMillerDev** [reply]
That's your issue. You have an ARM mac, but you added `/usr/local/opt/php/lib/httpd/modules/libphp.so` on line 182 of the httpd config. It should be `/opt/homebrew/opt/php/lib/httpd/modules/libphp.so` as you mentioned you added in your post.

**Reply by gromgit** [reply]
>```
>httpd: Syntax error on line 182 of /opt/homebrew/etc/httpd/httpd.conf: Cannot load /usr/local/opt/php/lib/httpd/modules/libphp.so into server: dlopen(/usr/local/opt/php/lib/httpd/modules/libphp.so, 0x000A): tried: '/usr/local/opt/php/lib/httpd/modules/libphp.so' (no such file)
>```

Are you trying to load a (nonexistent) Intel shared library into an Apple Silicon `httpd` binary? Because that doesn't make any sense.

**Reply by toloekka** [reply]
gromgit - maby I did a mistake.....

**Reply by gromgit** [reply]
Also, just noticed this from your original post:
>```
>LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "2msstudio"
>( I added my Home named «2msstudio» in .zprofile)
>```

What are you expecting to happen by adding `2msstudio` to the `LoadModule` directive? Because according to https://httpd.apache.org/docs/current/mod/mod_so.html#loadmodule, it's a useless argument.

**Reply by toloekka** [reply]
> Also, just noticed this from your original post:
> 
> > ```
> > LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so "2msstudio"
> > ( I added my Home named «2msstudio» in .zprofile)
> > ```
> 
> What are you expecting to happen by adding `2msstudio` to the `LoadModule` directive? Because according to https://httpd.apache.org/docs/current/mod/mod_so.html#loadmodule, it's a useless argument.

What I expected!? I expected it to work and to be necessary when I read this: https://www.codexpedia.com/apache-server/no-code-signing-authority-for-module-php-on-mac-monterey/
I did not try this out of the blue?

But now I tried the right path to load the module so now apache/httpd works.

**Reply by toloekka** [reply]
> That's your issue. You have an ARM mac, but you added `/usr/local/opt/php/lib/httpd/modules/libphp.so` on line 182 of the httpd config. It should be `/opt/homebrew/opt/php/lib/httpd/modules/libphp.so` as you mentioned you added in your post.

Now httpd seems to work. Thanks a lot :)

**Reply by gromgit** [reply]
> What I expected!? I expected it to work and to be necessary when I read this: https://www.codexpedia.com/apache-server/no-code-signing-authority-for-module-php-on-mac-monterey/

That article talks about codesigning for _Apple's_ `httpd` (it tells you to look for the server logs in `/var/log`). I'm not sure it's even relevant for Homebrew's `httpd`, and it definitely isn't documented in the standard Apache documentation

**Reply by toloekka** [reply]
> > What I expected!? I expected it to work and to be necessary when I read this: https://www.codexpedia.com/apache-server/no-code-signing-authority-for-module-php-on-mac-monterey/
> 
> That article talks about codesigning for _Apple's_ `httpd` (it tells you to look for the server logs in `/var/log`). I'm not sure it's even relevant for Homebrew's `httpd`, and it definitely isn't documented in the standard Apache documentation

I am definitely not sure if its relevant in Homebrew or not but when I get errors I Google and try different things and I tried "2msstudio" at the end of the LoadModule line. That did not affect if its loaded or not, but then I found out.

**Comment by toloekka** [comment]
Sorry folks :(

After starting my Mac this morning I got this error back again when I use "brew services list " (and now I have "LoadModule php_module /opt/homebrew/opt/php/lib/httpd/modules/libphp.so" in the httpd.conf file):
httpd   error  256 root      ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

The error_log say this:
[Wed Aug 17 07:27:39.607595 2022] [mpm_prefork:notice] [pid 120] AH00163: Apache/2.4.54 (Unix) PHP/8.1.9 configured -- resuming normal operations
[Wed Aug 17 07:27:39.607800 2022] [core:notice] [pid 120] AH00094: Command line: '/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND'

When I use the command "/opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND",  I got this output:
(48)Address already in use: AH00072: make_sock: could not bind to address [::]:80
(48)Address already in use: AH00072: make_sock: could not bind to address 0.0.0.0:80
no listening sockets available, shutting down
AH00015: Unable to open logs

**Reply by SMillerDev** [reply]
You have another server running on port 80. You'll have to stop that one before you can use the port for httpd

**Comment by toloekka** [comment]
How can I find and stop that?

I used this command earlier:
sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist 2>/dev/null

"httpd -v" and "which httpd" give this output now:
2msstudio@Toms-Mac-Studio ~ % httpd -v
Server version: Apache/2.4.54 (Unix)
Server built:   Jun 11 2022 03:23:16
2msstudio@Toms-Mac-Studio ~ % 
2msstudio@Toms-Mac-Studio ~ % which httpd
/opt/homebrew/bin/httpd
2msstudio@Toms-Mac-Studio ~ %

**Reply by SMillerDev** [reply]
How to find a running server on my machine is hardly a homebrew question. But I'll try anyways.

Have you tried checking for httpd processes using Activity Monitor?

**Reply by toloekka** [reply]
I found this in Acitivity Monitor:
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	385	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	997	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	386	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	387	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,02	1	0	Apple	0,0	0,00	1026	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,03	1	0	Apple	0,0	0,00	383	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	1027	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)	
 httpd	0,0	0,01	1	0	Apple	0,0	0,00	384	2msstudio		0 byte	0 byte	0 byte	Nei	Nei	Nei	(null)	Nei	0 byte	0	0 byte	0	0 byte	0 byte	0 byte	-	0 byte	0 byte	Nei	(null)

**Reply by toloekka** [reply]
I did this commands again to stop Apples built in httpd:
2msstudio@Toms-Mac-Studio ~ % sudo apachectl stop
Password:
2msstudio@Toms-Mac-Studio ~ % sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist 2>/dev/null
2msstudio@Toms-Mac-Studio ~ % 

and after restating "brew services restart httpd", the httpd works again:
2msstudio@Toms-Mac-Studio ~ % brew services list 
Name    Status  User      File
httpd   started root      ~/Library/LaunchAgents/homebrew.mxcl.httpd.plist

I thought that this command "sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist 2>/dev/null" would stop the Apple built in httpd "for ever".
